src/main/java/net/dries007/tfc/common/container/KnappingContainer.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Licensed under the EUPL, Version 1.2.
+ * You may obtain a copy of the Licence at:
+ * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
+ */
+
+package net.dries007.tfc.common.container;
+
+import javax.annotation.Nullable;
+
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.PlayerInventory;
+import net.minecraft.inventory.container.ContainerType;
+import net.minecraft.inventory.container.Slot;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.IRecipeType;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.util.Hand;
+import net.minecraft.util.SoundEvent;
+import net.minecraft.world.server.ServerWorld;
+
+import net.minecraftforge.items.ItemHandlerHelper;
+import net.minecraftforge.items.ItemStackHandler;
+
+import net.dries007.tfc.common.TFCTags;
+import net.dries007.tfc.common.recipes.IInventoryNoop;
+import net.dries007.tfc.common.recipes.SimpleCraftMatrix;
+import net.dries007.tfc.common.recipes.KnappingRecipe;
+import net.dries007.tfc.util.Helpers;
+
+public class KnappingContainer extends ItemStackContainer implements IButtonHandler, IInventoryNoop
+{
+    public final SimpleCraftMatrix matrix;
+    private final int amountToConsume;
+    public final boolean usesDisabledTex;
+    private final boolean needsKnife;
+    private final boolean consumeAfterComplete;
+    public final SoundEvent sound;
+    public boolean requiresReset;
+    private boolean hasBeenModified;
+    private boolean hasConsumedIngredient;
+    private final IRecipeType<? extends KnappingRecipe> recipeType;
+    public final ItemStack stackCopy;
+
+    public KnappingContainer(ContainerType<?> containerType, IRecipeType<? extends KnappingRecipe> recipeType, int windowId, PlayerInventory playerInv, int amountToConsume, boolean consumeAfterComplete, boolean usesDisabledTex, boolean needsKnife, SoundEvent sound)
+    {
+        super(containerType, windowId, playerInv, getItemForKnapping(playerInv), 20);
+        this.itemIndex += 1;
+        this.amountToConsume = amountToConsume;
+        this.usesDisabledTex = usesDisabledTex;
+        this.consumeAfterComplete = consumeAfterComplete;
+        this.needsKnife = needsKnife;
+        this.recipeType = recipeType;
+        this.sound = sound;
+
+        matrix = new SimpleCraftMatrix();
+        hasBeenModified = false;
+        requiresReset = false;
+        hasConsumedIngredient = false;
+        stackCopy = this.stack.copy();
+    }
+
+    private static ItemStack getItemForKnapping(PlayerInventory inv)
```

Comment:
Does this not break if you open a knapping gui with a non-empty mainhand but the knapping item in offhand?

Reply:
Yes it breaks